src/dspaces-server.c

```diff
@@ -1297,6 +1304,37 @@ static void put_rpc(hg_handle_t handle)
     // set the owner to be this server address
     odsc_take_ownership(server, &in_odsc);
 
+    /* Check if needs to swap data out to HDF5*/
+    uint64_t size_MB = obj_data_size(&in_odsc) >> 20;
+    struct obj_data_ptr_flat_list_entry *swap_od_entry;
+    struct obj_data *swap_od;
+
+    while(need_swap_out(&server->conf.swap, size_MB)) {
+        /* Find the od that we want to swap out */
+        ABT_mutex_lock(server->ls_mutex);
+        swap_od = which_swap_out(&server->conf.swap, &server->dsg->ls_od_list);
+        ABT_mutex_unlock(server->ls_mutex);
```

Comment:
You're right. I'll make the critical section cover the `search()` and `remove()` at this moment. But the intention that I split the critical section here is to prevent the thread holding the lock during the time-consuming I/O interval. Perhaps I can set a flag here and ask the `search()` to find the next `od` if the one it finds has the flag set. Then we can switch back to this implementation.

Reply:
You could delete it from local storage before you write it to disk.

Reply:
I keep the od exists until the write finishes because I'm afraid the write fails and we lose the od. But maybe I can delete it first, if the write fails, then we add it back.